californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -234,6 +235,13 @@ public Exchange receiveResponse(final Response response, final CorrelationContex
 			return null;
 		} else if (correlationContextMatcher.isResponseRelatedToRequest(exchange.getCorrelationContext(), responseContext)) {
 
+			if (response.getType() == Type.ACK && exchange.getCurrentRequest().getMID() != response.getMID()) {
```

Comment:
An ACK is in the first place matched by the MID. Only for "piggy-backed" ACKs, the token may also be used. The current implementation inverts this and therefore cause this erroneous behaviour.

The processing of other responses (CON/NON) is not changed by this PR.